distributed/client.py

```diff
@@ -1621,10 +1621,7 @@ def map(
         total_length = sum(len(x) for x in iterables)
 
         if batch_size and batch_size > 1 and total_length > batch_size:
-            # PREM
-            batches = list(
-                zip(*[partition_all(batch_size, iterable) for iterable in iterables])
-            )
+            batches = zip(*[partition_all(batch_size, iterable) for iterable in iterables])
```

Comment:
This creates a temporary list just to unpack it? Could we just create a zip iterator from a generator expression here to avoid materializing an intermediate list unnecessarily?

Reply:
@kkraus14 So the actual materialization is removed with the removal of the outer `list`. Now there is no materialization with the usage of inner `[]`, we will need it to create `partition_all` objects list(like.. `[partition_all, partition_all,...]`)

Also, there seems to be a need for `zip` in this particular logic as to me it looks like the iterables are being expanded into repetitions.

Reply:
You're creating an intermediate list via list comprehension in `[partition_all(batch_size, iterable) for iterable in iterables]` which you then unpack into the zip iterator. I'm saying you should be able to just use a generator to avoid materializing that list:
`zip(*(partition_all(batch_size, iterable) for iterable in iterables))`

Reply:
I think that we may be at the point of diminishing returns here.  If this isn't happening within a tight for loop then it probably doesn't matter.

There is also some debuggability cost to using generators over lists.  The Python debugger doesn't descend as well well when we unpack generators than when we unpack lists.